lib/kuzzle/internalIndexHandler.js

```diff
@@ -0,0 +1,241 @@
+      if (error instanceof Bluebird.TimeoutError) {
+        throw kerror.get('services', 'storage', 'bootstrap_timeout', 'internalIndex');
+      }
+
+      throw error;
+    }
+    finally {
+      await mutex.unlock();
+    }
+  }
+
+  /**
+   * @override
+   */
+  async _bootstrapSequence () {
+    debug('Bootstrapping security structure');
+    await this.createInitialSecurities();
+
+    debug('Bootstrapping document validation structure');
+    await this.createInitialValidations();
+
+    debug('Bootstrapping JWT secret');
+    await this._persistSecret();
+
+    debug('Loading API keys into Redis');
+    await this._loadApiKeys();
+
+    // Create datamodel version
+    await this.create('config', { version: dataModelVersion }, {
+      id: this._DATAMODEL_VERSION_ID,
+    });
+  }
+
+  /**
+   * Creates initial roles and profiles as specified in Kuzzle configuration
+   */
+  async createInitialSecurities () {
+    await Bluebird.map(
+      Object.entries(securitiesBootstrap.roles),
+      ([roleId, content]) => {
+        return this.createOrReplace('roles', roleId, content, {
+          refresh: 'wait_for',
+        });
+      });
+
+    await Bluebird.map(
+      Object.entries(securitiesBootstrap.profiles),
+      ([profileId, content]) => {
+        return this.createOrReplace('profiles', profileId, content, {
+          refresh: 'wait_for',
+        });
+      });
+  }
+
+  async createInitialValidations () {
+    const initialValidations = this.kuzzle.config.validation;
+    const promises = [];
+
+    for (const [index, collection] of Object.entries(initialValidations)) {
+      for (const [collectionName, validation] of Object.entries(collection)) {
+        const validationId = `${index}#${collectionName}`;
+
+        promises.push(
+          this.createOrReplace('validations', validationId, validation));
+      }
+    }
+
+    await Bluebird.all(promises);
+  }
+
+  async getSecret () {
+    const response = await this.get('config', this._JWT_SECRET_ID);
+
+    return response._source.seed;
+  }
+
+  async _persistSecret () {
+    const seed = this.kuzzle.config.security.jwt.secret
+      || crypto.randomBytes(512).toString('hex');
+
+    try {
+      await this.create('config', { seed }, {
+        id: this._JWT_SECRET_ID,
+      });
+    }
+    catch (e) {
+      if (e.id !== 'services.storage.document_already_exists') {
+        throw e;
+      }
+
+      // ignore: the secret already exists
+    }
+  }
+
+  /**
+   * Loads authentication token from API key into Redis
+   *
+   * @returns {Promise}
+   */
+  async _loadApiKeys () {
+    const promises = [];
+    const createToken = obj => this.kuzzle.ask(
+      'core:security:token:assign',
+      obj.token,
+      obj.userId,
+      obj.ttl);
+
+    await ApiKey.batchExecute({ match_all: {} }, apiKeys => {
+      for (const { _source } of apiKeys) {
+        promises.push(createToken(_source));
+      }
+    });
+
+    return Bluebird.all(promises);
```

Comment:
We should put a concurrency limit here in case there is a lot of API keys

Reply:
You might be right here. We should do that in another PR though: this requires rewriting the BaseModel's batchExecute logic, and I think that this is outside the scope of this PR.

Reply:
Oh yes you're right.